jnipp.h

```diff
@@ -70,11 +71,73 @@ namespace jni
     typedef std::runtime_error Exception;
 
 #endif // JNIPP_EXCEPTION_CLASS
+#endif // !JNIPP_DISABLE_EXCEPTIONS
 
     // Foward Declarations
     class Object;
     template <class TElement> class Array;
 
+    enum class ExceptionCategory {
+      Initialization,
+      //! A supplied name or signature could not be found
+      NameResolution,
+      Invocation,
+    };
+
+    /// Class for exception data, used for noexcept overloads
+    class ExceptionData {
+      public:
+        ExceptionData() noexcept = default;
+        ~ExceptionData() noexcept;
+        /// move constructor
+        ExceptionData(ExceptionData &&other) noexcept;
+        /// move assignment
+        ExceptionData &operator=(ExceptionData &&other) noexcept;
+        // no copy
+        ExceptionData(const ExceptionData &) = delete;
+        ExceptionData& operator=(const ExceptionData &) = delete;
+
+        explicit operator bool() const noexcept { return exceptionThrown; }
+
+        /// Assign from a category and string with static lifetime
+        void setFromStaticMessage(ExceptionCategory cat,
+                                  const char *msg) noexcept;
+
+        void setFromDynamicMessage(ExceptionCategory cat,
+                                  std::string&& msg) noexcept;
+
+        /// Assign from a category and JNI exception
+        void setFromJni(ExceptionCategory cat, JNIEnv *env,
+                        jobject exception) noexcept;
+
+        /// If an exception is set in the JNIEnv, get it, and clear it from the env.
+        void setFromJniEnv(ExceptionCategory cat, JNIEnv* env) noexcept;
+
+        /// Reset the contents
+        void reset() noexcept;
+
+        /// Swap with another instance of ExceptionData
+        void swap(ExceptionData &other) noexcept;
+
+#ifndef JNIPP_DISABLE_EXCEPTIONS
+        /// Throw an Exception subclass based on this data, if set as thrown.
+        void throwException();
+#endif // !JNIPP_DISABLE_EXCEPTIONS
+
+        /// Get a string from the exception, if any.
+        std::string getMessage() const;
+
+      private:
+        std::string getExceptionObjectMessage() const;
+        bool exceptionThrown = false;
+        ExceptionCategory category{};
+        const char *staticMessage = nullptr;
```

Comment:
Nitpick: I think this approach is error-prone - IMO it should be OK to just have a `std::string` member that'll get populated when some exception is being thrown. For one, IDK if there's a way to enforce that `setFromStaticMessage()` is actually called with a string literal. Another problem is that this class needs to correctly handle each of the 3 cases (static message, dynamic message, exception object) and I think `dynamicMessage` is not currently handled correctly in `throwException()` & in `getMessage()`.

Reply:
yeah, dynamic message was kind of half-added. I wasn't 100% sure what class of exceptions we should be robust to, so I was trying to avoid non-"noexcept" functions including the string constructor if possible.

(TBH I rarely work with exceptions disabled, so I have no real good mental model of how C++ works without that functionality.)

Reply:
Hmm, I didn't think of that... I see that `std::runtime_error`'s ctors aren't noexcept so IDK how important it is to stick to such methods here. I think they also will have to make a copy of the data for the same reasons we'd have to copy it here (i.e. `const char*` doesn't guarantee that the ptr will be forever valid)? libc++'s [implementation](https://github.com/llvm/llvm-project/blob/e15fcd7ba41e63edc7da217f0b33aeb10f1c0b5a/libcxx/src/include/refstring.h#L85) seems to introduce refcounting here as well, but in the end there's going to be a `new` + `memcpy` (IIUC this is called from runtime_error's [ctor](https://github.com/llvm/llvm-project/blob/e15fcd7ba41e63edc7da217f0b33aeb10f1c0b5a/libcxx/src/support/runtime/stdexcept_default.ipp#L38)). It may be that `std::string` ctor and `std::runtime_error` ctor cannot be noexcept because allocation can always throw, so maybe we don't have to worry about this?